File: mobile/lib/features/workout/domain/exercise.dart
// =====================================================================
// MiCoach — Catálogo de ejercicios (mirror de WorkoutDtos en backend).
// =====================================================================
class ExerciseMuscle {
  final int muscleId;
  final String muscleCode;
  final String muscleName;
  final String role;

  const ExerciseMuscle(
      {required this.muscleId, required this.muscleCode, required this.muscleName, required this.role});

  factory ExerciseMuscle.fromJson(Map<String, dynamic> json) => ExerciseMuscle(
        muscleId: json['muscleId'] as int,
        muscleCode: json['muscleCode'] as String,
        muscleName: json['muscleName'] as String,
        role: json['role'] as String,
      );
}

class Exercise {
  final int id;
  final String name;
  final String? description;
  final String category;
  final List<String> equipment;
  final String difficulty;
  final String? instructions;
  final String? videoUrl;
  final String? imageUrl;
  /// 'reps' (default) o 'duration' — ej. Plancha se sostiene X segundos, no se repite X veces.
  final String measurementType;
  final bool aiGenerated;
  final List<ExerciseMuscle> muscles;

  const Exercise({
    required this.id,
    required this.name,
    this.description,
    required this.category,
    this.equipment = const [],
    required this.difficulty,
    this.instructions,
    this.videoUrl,
    this.imageUrl,
    this.measurementType = 'reps',
    this.aiGenerated = false,
    this.muscles = const [],
  });

  factory Exercise.fromJson(Map<String, dynamic> json) => Exercise(
        id: json['id'] as int,
        name: json['name'] as String,
        description: json['description'] as String?,
        category: json['category'] as String,
        equipment: (json['equipment'] as List<dynamic>? ?? const []).map((e) => e as String).toList(),
        difficulty: json['difficulty'] as String,
        instructions: json['instructions'] as String?,
        videoUrl: json['videoUrl'] as String?,
        imageUrl: json['imageUrl'] as String?,
        measurementType: json['measurementType'] as String? ?? 'reps',
        aiGenerated: json['aiGenerated'] as bool? ?? false,
        muscles: (json['muscles'] as List<dynamic>? ?? const [])
            .map((e) => ExerciseMuscle.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/modules/workout/src/main/java/com/micoach/workout/domain/Exercise.java
package com.micoach.workout.domain;

import lombok.Getter;

import java.time.Instant;
import java.util.List;

/**
 * Catálogo de ejercicios (tabla workout_exercises), con sus músculos involucrados.
 */
@Getter
public class Exercise {

    private final Long id;
    private final String name;
    private final String description;
    private final String category;
    private final List<String> equipment;
    private final String difficulty;
    private final String instructions;
    private final String videoUrl;
    private final String imageUrl;
    /** Posición final/contraída del movimiento (docs/10-recomendaciones-coach-nutricion.md § H.6). */
    private final String imageUrlEnd;
    /** 'reps' (por defecto) o 'duration' — ej. Plancha se sostiene X segundos, no se repite X veces. */
    private final String measurementType;
    private final boolean aiGenerated;
    private final boolean active;
    private final List<ExerciseMuscle> muscles;
    private final Instant createdAt;
    private final Instant updatedAt;

    private Exercise(Long id, String name, String description, String category,
                     List<String> equipment, String difficulty, String instructions,
                     String videoUrl, String imageUrl, String imageUrlEnd, String measurementType,
                     boolean aiGenerated, boolean active, List<ExerciseMuscle> muscles, Instant createdAt,
                     Instant updatedAt) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.category = category;
        this.equipment = equipment == null ? List.of() : equipment;
        this.difficulty = difficulty;
        this.instructions = instructions;
        this.videoUrl = videoUrl;
        this.imageUrl = imageUrl;
        this.imageUrlEnd = imageUrlEnd;
        this.measurementType = measurementType == null ? "reps" : measurementType;
        this.aiGenerated = aiGenerated;
        this.active = active;
        this.muscles = muscles == null ? List.of() : muscles;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static Exercise restore(Long id, String name, String description, String category,
                                   List<String> equipment, String difficulty, String instructions,
                                   String videoUrl, String imageUrl, String imageUrlEnd, String measurementType,
                                   boolean aiGenerated, boolean active, List<ExerciseMuscle> muscles,
                                   Instant createdAt, Instant updatedAt) {
        return new Exercise(id, name, description, category, equipment, difficulty, instructions,
                videoUrl, imageUrl, imageUrlEnd, measurementType, aiGenerated, active, muscles, createdAt,
                updatedAt);
    }
}
